File: api/agent.py
"""Blueprints de features agénticas para Lofi Pomodoro.

Endpoints (todos POST /api/agent/...):
  - /split     -> descompone una tarea en micro-pasos ~10 min
  - /plan      -> top 3 de foco para el día
  - /ritual    -> coach del shutdown ritual
  - /insights  -> patrones semanales + tareas atascadas
  - /health    -> disponibilidad del agente (sin llamar al LLM)
"""

import traceback

from flask import Blueprint, jsonify, request

agent_bp = Blueprint('agent', __name__)


def _agent_or_error():
    """Devuelve (client, error_response|None)."""
    try:
        from agent_client import get_agent_client
        return get_agent_client(), None
    except Exception as e:
        print(f"[agent] unavailable: {e}")
        return None, (jsonify({
            'error': 'El asistente IA no está disponible. Configurá NVIDIA_API_KEY en .env',
        }), 503)


def _ok(payload):
    return jsonify({'ok': True, **payload}), 200


def _fail(msg, code=200):
    return jsonify({'ok': False, 'error': msg}), code


# --------------------------------------------------------------------- #
#  Split de tareas
# --------------------------------------------------------------------- #
@agent_bp.route('/split', methods=['POST'])
def split():
    data = request.get_json(silent=True) or {}
    title = (data.get('title') or '').strip()
    if not title:
        return _fail('Falta el título de la tarea', 400)

    client, err = _agent_or_error()
    if err:
        return err
    try:
        result = client.split_task(
            title=title,
            energy=data.get('energy', 'medium'),
            context=(data.get('context') or '').strip(),
            profile=data.get('profile'),
        )
        return _ok({'parts': result['parts']})
    except Exception as e:
        traceback.print_exc()
        return _fail(f'No pude dividir la tarea: {e}')


# --------------------------------------------------------------------- #
#  Plan del día
# --------------------------------------------------------------------- #
@agent_bp.route('/plan', methods=['POST'])
def plan():
    data = request.get_json(silent=True) or {}
    if not isinstance(data.get('tasks'), list):
        return _fail('Falta la lista de tareas', 400)

    client, err = _agent_or_error()
    if err:
        return err
    try:
        result = client.daily_plan(
            pending=data.get('tasks') or [],
            energy_today=data.get('energyToday') or 'medium',
            streak=int(data.get('streak') or 0),
            note=data.get('note') or '',
            yesterday_focus=data.get('yesterdayFocus') or '',
            profile=data.get('profile'),
        )
        return _ok(result)
    except Exception as e:
        traceback.print_exc()
        return _fail(f'No pude armar el plan: {e}')


# --------------------------------------------------------------------- #
#  Coach del ritual de cierre
# --------------------------------------------------------------------- #
@agent_bp.route('/ritual', methods=['POST'])
def ritual():
    data = request.get_json(silent=True) or {}

    client, err = _agent_or_error()
    if err:
        return err
    try:
        result = client.ritual_reflection(
            went_well=data.get('wentWell') or '',
            was_hard=data.get('wasHard') or '',
            tomorrow_focus=data.get('tomorrowFocus') or '',
            mood=data.get('mood') or 'medium',
            completed_count=int(data.get('completedCount') or 0),
            streak=int(data.get('streak') or 0),
            profile=data.get('profile'),
        )
        return _ok(result)
    except Exception as e:
        traceback.print_exc()
        return _fail(f'No pude generar la reflexión: {e}')


# --------------------------------------------------------------------- #
#  Insights semanales + tareas atascadas
# --------------------------------------------------------------------- #
@agent_bp.route('/insights', methods=['POST'])
def insights():
    data = request.get_json(silent=True) or {}
    if not isinstance(data.get('tasks'), list):
        return _fail('Falta la lista de tareas', 400)

    client, err = _agent_or_error()
    if err:
        return err
    try:
        result = client.weekly_insights(
            tasks=data.get('tasks') or [],
            shutdown_entries=data.get('shutdownEntries') or [],
            completed_counts=data.get('completedCounts') or [],
            streak=int(data.get('streak') or 0),
            profile=data.get('profile'),
        )
        return _ok(result)
    except Exception as e:
        traceback.print_exc()
        return _fail(f'No pude generar los insights: {e}')


# --------------------------------------------------------------------- #
#  Health (sin llamar al LLM)
# --------------------------------------------------------------------- #
@agent_bp.route('/health', methods=['GET'])
def health():
    try:
        from agent_client import get_agent_client
        get_agent_client()
        return _ok({'available': True})
    except Exception:
        return _ok({'available': False})

File: frontend/src/lib/agent.ts
/**
 * Cliente tipado de las features agénticas (api/agent.py).
 *
 * Todas las respuestas llegan con { ok: true, ... } o { ok: false, error }.
 */
import { post, get } from './api';
import type { Energy, Task, ShutdownEntry } from './stores/singleton';
import { profilePayload } from './stores/profile.svelte';

export type AgentPart = { title: string; estimatedMinutes: number };

export type PlanFocus = {
  taskId: string | null;
  title: string;
  reason: string;
};

export type DailyPlan = {
  message: string;
  focus: PlanFocus[];
};

export type RitualReflection = {
  reflection: string;
  tip: string;
};

export type StuckTask = {
  title: string;
  daysInDoing: number;
  taskId?: string;
  suggestion: string;
};

export type WeeklyInsights = {
  summary: string;
  insights: string[];
  stuckTasks: StuckTask[];
};

type AgentResponse<T> = { ok: true } & T;

/** Procesa la respuesta agéntica lanzando Error si el backend reportó fallo. */
function unwrap<T>(res: { ok: boolean; error?: string } & T): T {
  if (!res.ok) throw new Error(res.error ?? 'El asistente no respondió');
  return res;
}

export async function agentSplit(
  title: string,
  energy: Energy = 'medium',
  context = '',
): Promise<{ parts: AgentPart[] }> {
  const res = await post<AgentResponse<{ parts: AgentPart[] }>>('/api/agent/split', {
    title,
    energy,
    context,
    profile: profilePayload(),
  });
  return unwrap(res);
}

export async function agentPlan(args: {
  tasks: Task[];
  energyToday?: Energy | 'all';
  streak?: number;
  note?: string;
  yesterdayFocus?: string;
}): Promise<DailyPlan> {
  const res = await post<AgentResponse<DailyPlan>>('/api/agent/plan', {
    tasks: args.tasks,
    energyToday: args.energyToday ?? 'medium',
    streak: args.streak ?? 0,
    note: args.note ?? '',
    yesterdayFocus: args.yesterdayFocus ?? '',
    profile: profilePayload(),
  });
  return unwrap(res);
}

export async function agentRitual(args: {
  wentWell: string;
  wasHard: string;
  tomorrowFocus: string;
  mood: Energy;
  completedCount: number;
  streak: number;
}): Promise<RitualReflection> {
  const res = await post<AgentResponse<RitualReflection>>('/api/agent/ritual', {
    ...args,
    profile: profilePayload(),
  });
  return unwrap(res);
}

export async function agentInsights(args: {
  tasks: Task[];
  shutdownEntries: ShutdownEntry[];
  completedCounts: { date: string; count: number }[];
  streak: number;
}): Promise<WeeklyInsights> {
  const res = await post<AgentResponse<WeeklyInsights>>('/api/agent/insights', {
    tasks: args.tasks,
    shutdownEntries: args.shutdownEntries,
    completedCounts: args.completedCounts,
    streak: args.streak,
    profile: profilePayload(),
  });
  return unwrap(res);
}

export async function agentHealth(): Promise<{ available: boolean }> {
  const res = await get<AgentResponse<{ available: boolean }>>('/api/agent/health');
  return unwrap(res);
}